File: 2024/day-04.rb
#!/usr/bin/env ruby

# https://adventofcode.com/2024/day/4

require "bundler/setup"
Bundler.require

matrix = File.readlines("#{__dir__}/day-04.txt", chomp: true).map(&:chars)

height = matrix.size
width = matrix.first.size

xmas = "XMAS".chars
rxmas = xmas.reverse

part1 = width.times.to_a
  .product(height.times.to_a)
  .reduce(0) do |acc, (x, y)|
    [
      [0, 0,   1, 0], # Horizontal
      [0, 0,   1, 1], # Diagonal TL -> BR
      [0, 0,   0, 1], # Vertical
      [3, 0,  -1, 1]  # Diagonal TR -> BL
    ].reduce(acc) do |wacc, (ox, oy, dx, dy)|
      window = xmas.size.times.map { |i| matrix.at(y + oy + i * dy)&.at(x + ox + i * dx) }
      (window == xmas || window == rxmas) ? wacc + 1 : wacc
    end
  end

puts part1

part2 = (1...width - 1).to_a
  .product((1...height - 1).to_a)
  .reduce(0) do |acc, (x, y)|
    unless matrix[y][x] == "A"
      next acc
    end

    tl = matrix[y - 1][x - 1]
    tr = matrix[y - 1][x + 1]
    bl = matrix[y + 1][x - 1]
    br = matrix[y + 1][x + 1]

    [tl, tr, bl, br].sort == %w[M M S S] && tl != br ? acc + 1 : acc
  end

puts part2

File: 2024/day-04.exs
#!/usr/bin/env elixir

# https://adventofcode.com/2024/day/4

defmodule Point do
  def create(x, y), do: {x, y}

  def add({x1, y1}, {x2, y2}), do: {x1 + x2, y1 + y2}
end

matrix =
  File.stream!("day-04.txt")
  |> Enum.map(&String.trim/1)
  |> Enum.map(&String.graphemes/1)

directions = [
  {0, 0, 1, 0},
  {0, 0, 1, 1},
  {0, 0, 0, 1},
  {3, 0, -1, 1}
]

xmas = String.graphemes("XMAS")
rxmas = String.graphemes("SAMX")

matrix_element = fn matrix, {x, y} ->
  case Enum.at(matrix, y, nil) do
    nil ->
      nil

    row ->
      Enum.at(row, x, nil)
  end
end

matrix_elements = fn matrix, {x, y}, {ox, oy, dx, dy}, count ->
  0..(count - 1)
  |> Enum.map(fn i ->
    matrix_element.(matrix, {x + ox + dx * i, y + oy + dy * i})
  end)
end

part1 =
  Enum.reduce(Enum.with_index(matrix), 0, fn {row, y}, row_acc ->
    Enum.reduce(0..(Enum.count(row) - 1), row_acc, fn x, col_acc ->
      Enum.reduce(directions, col_acc, fn direction, acc ->
        elements = matrix_elements.(matrix, {x, y}, direction, Enum.count(xmas))

        if elements == xmas || elements == rxmas do
          acc + 1
        else
          acc
        end
      end)
    end)
  end)

IO.inspect(part1)

part2 =
  Enum.reduce(Enum.with_index(matrix), 0, fn {row, y}, row_acc ->
    Enum.reduce(Enum.with_index(row), row_acc, fn {el, x}, acc ->
      case el do
        "A" ->
          tl = matrix_element.(matrix, {x - 1, y - 1})
          tr = matrix_element.(matrix, {x + 1, y - 1})
          bl = matrix_element.(matrix, {x - 1, y + 1})
          br = matrix_element.(matrix, {x + 1, y + 1})

          sorted = Enum.sort([tl, tr, bl, br])

          if sorted === String.graphemes("MMSS") && tl != br do
            acc + 1
          else
            acc
          end

        _ ->
          acc
      end
    end)
  end)

IO.inspect(part2)
